File: scripts/generate-hero.mjs
import { readFileSync, writeFileSync } from "node:fs";

const width = 1024;
const height = 1024;
const purple = "#5d479a";
const white = "#f5feff";
const banana = readFileSync("public/images/bananas.png").toString("base64");

const glyphs = {
  A: ["01110", "10001", "10001", "11111", "10001", "10001", "10001"],
  B: ["11110", "10001", "10001", "11110", "10001", "10001", "11110"],
  E: ["11111", "10000", "10000", "11110", "10000", "10000", "11111"],
  F: ["11111", "10000", "10000", "11110", "10000", "10000", "10000"],
  M: ["10001", "11011", "10101", "10101", "10001", "10001", "10001"],
  N: ["10001", "11001", "10101", "10011", "10001", "10001", "10001"],
  P: ["11110", "10001", "10001", "11110", "10000", "10000", "10000"],
  R: ["11110", "10001", "10001", "11110", "10100", "10010", "10001"],
  T: ["11111", "00100", "00100", "00100", "00100", "00100", "00100"],
  Y: ["10001", "10001", "01010", "00100", "00100", "00100", "00100"],
  "0": ["01110", "10001", "10011", "10101", "11001", "10001", "01110"],
  "3": ["11110", "00001", "00001", "01110", "00001", "00001", "11110"],
  "4": ["10010", "10010", "10010", "11111", "00010", "00010", "00010"],
  "9": ["01110", "10001", "10001", "01111", "00001", "00001", "01110"],
  ".": ["00000", "00000", "00000", "00000", "00000", "01100", "01100"],
  "@": ["01110", "10001", "10111", "10101", "10111", "10000", "01110"],
  "↑": ["00100", "01110", "10101", "00100", "00100", "00100", "00100"],
  " ": ["00000", "00000", "00000", "00000", "00000", "00000", "00000"],
};

const smallGlyphs = {
  ...glyphs,
  a: ["00000", "00000", "01110", "00001", "01111", "10001", "01111"],
  b: ["00000", "10000", "10000", "11110", "10001", "10001", "11110"],
  e: ["00000", "00000", "01110", "10001", "11111", "10000", "01110"],
  h: ["10000", "10000", "10000", "11110", "10001", "10001", "10001"],
  m: ["00000", "00000", "11010", "10101", "10101", "10101", "10101"],
  o: ["00000", "00000", "01110", "10001", "10001", "10001", "01110"],
  r: ["00000", "00000", "10110", "11001", "10000", "10000", "10000"],
  t: ["00100", "00100", "11111", "00100", "00100", "00101", "00010"],
  x: ["00000", "00000", "10001", "01010", "00100", "01010", "10001"],
  y: ["00000", "00000", "10001", "10001", "01111", "00001", "01110"],
};

function pixelText(text, x, y, scale, source = glyphs) {
  const spacing = scale;
  const charWidth = 5 * scale + spacing;
  let cursor = x;
  let rects = "";

  for (const char of text) {
    const grid = source[char] ?? source[" "];
    for (let row = 0; row < grid.length; row++) {
      for (let col = 0; col < grid[row].length; col++) {
        if (grid[row][col] === "1") {
          rects += `<rect x="${cursor + col * scale}" y="${y + row * scale}" width="${scale}" height="${scale}" fill="${white}"/>`;
        }
      }
    }
    cursor += charWidth;
  }

  return rects;
}

function textWidth(text, scale) {
  return text.length * (5 * scale + scale) - scale;
}

function centered(text, y, scale, source = glyphs) {
  return pixelText(text, (width - textWidth(text, scale)) / 2, y, scale, source);
}

const svg = `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}" viewBox="0 0 ${width} ${height}">
  <rect width="100%" height="100%" fill="${purple}"/>
  ${centered("BANANA", 150, 13)}
  <image href="data:image/png;base64,${banana}" x="251" y="260" width="520" height="346" preserveAspectRatio="xMidYMid meet"/>
  ${centered("TAP ↑", 725, 13)}
  ${centered("Frame by @0x94t3z.eth", 880, 6, smallGlyphs)}
</svg>`;

writeFileSync("/tmp/banana-hero.svg", svg);

File: scripts/generate-hero.py
from pathlib import Path

from PIL import Image, ImageDraw, ImageFont

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "public/images/banana-hero.png"
BANANA = ROOT / "public/images/bananas.png"
FONT_400 = ROOT / "assets/fonts/pixelify-sans-400.ttf"
FONT_600 = ROOT / "assets/fonts/pixelify-sans-600.ttf"

W = 1024
H = 1024
PURPLE = (93, 71, 154)
WHITE = (245, 254, 255)


def draw_centered(draw: ImageDraw.ImageDraw, text: str, y: int, font: ImageFont.FreeTypeFont, spacing: int = 0) -> None:
    bbox = draw.textbbox((0, 0), text, font=font)
    text_w = bbox[2] - bbox[0]
    x = (W - text_w) // 2
    draw.text((x, y), text, font=font, fill=WHITE, spacing=spacing)


def draw_up_arrow(draw: ImageDraw.ImageDraw, x: int, y: int, scale: int) -> None:
    pixels = [
        "00100",
        "01110",
        "10101",
        "00100",
        "00100",
        "00100",
        "00100",
    ]
    for row, line in enumerate(pixels):
        for col, value in enumerate(line):
            if value == "1":
                draw.rectangle(
                    [
                        x + col * scale,
                        y + row * scale,
                        x + (col + 1) * scale - 1,
                        y + (row + 1) * scale - 1,
                    ],
                    fill=WHITE,
                )


def draw_tap_with_arrow(draw: ImageDraw.ImageDraw, y: int, font: ImageFont.FreeTypeFont) -> None:
    text = "TAP"
    text_bbox = draw.textbbox((0, 0), text, font=font)
    text_w = text_bbox[2] - text_bbox[0]
    arrow_scale = 10
    arrow_w = 5 * arrow_scale
    gap = 64
    total_w = text_w + gap + arrow_w
    text_x = (W - total_w) // 2
    arrow_x = text_x + text_w + gap
    draw.text((text_x, y), text, font=font, fill=WHITE)
    draw_up_arrow(draw, arrow_x, y + 16, arrow_scale)


def main() -> None:
    canvas = Image.new("RGBA", (W, H), (*PURPLE, 255))
    draw = ImageDraw.Draw(canvas)

    title_font = ImageFont.truetype(str(FONT_600), 82)
    tap_font = ImageFont.truetype(str(FONT_600), 82)
    credit_font = ImageFont.truetype(str(FONT_600), 38)

    banana = Image.open(BANANA).convert("RGBA")
    banana.thumbnail((520, 360), Image.Resampling.NEAREST)
    banana_x = (W - banana.width) // 2
    banana_y = 265

    draw_centered(draw, "BANANA", 145, title_font)
    canvas.alpha_composite(banana, (banana_x, banana_y))
    draw_tap_with_arrow(draw, 720, tap_font)
    draw_centered(draw, "Snap by @0x94t3z.eth", 875, credit_font)

    canvas.convert("RGB").save(OUT)


if __name__ == "__main__":
    main()
